features/0351-generic-json-message/README.md

```diff
@@ -0,0 +1,114 @@
+
+## Tutorial
+
+#### Name and Version
+
+This is the generic JSON message protocol, defined uniquely by the URI:
+
+    "https://didcomm.org/json/1.0"
+
+#### Key Concepts
+
+> This RFC assumes familiarity with [mediators and relays](/concepts/0046-mediators-and-relays/README.md), [attachments](/concepts/0017-attachments/README.md), and [message threading](/concepts/0008-message-id-and-threading/README.md)
+
+This protocol has two primary roles, the **sender** and the **receiver**. The sender constructs a message with a JSON payload, which the receiver parses. This protocol supports optional threading, so the receiver can send a response to the sender. 
+
+Secondary roles/sub-roles: In a use-case where the *sender* is not a DIDComm-enabled agent, we introduce the role of **sender-agent**, which is the agent that relays messages to/from the sender, translating from DIDComm to non-DIDComm messaging. Similarly, for a *receiver* that is not a DIDComm-enabled agent, we introduce the role of **receiver-agent**, which serves the same purpose for the receiver. 
+
+Note: if we refer to a sender-agent or receiver-agent, and an implementation doesn't have one, the relevant behaviour would instead be in the sender/receiver.
+
+With JSON messaging, custom agents can communicate domain-specific information or implement domain-specific protocols without breaking compatibility with the wider DIDComm standards. Agents can also [relay](/concepts/0046-mediators-and-relays/README.md) messages between entities that are not DIDComm-enabled, allowing a wide range of communication protocols to use DIDComm to transport their messages.
+
+This protocol does not define the format of the content of a message, except that it be an [embedded attachment](/concepts/0017-attachments/README.md). Specific implementations may specify the exact format they expect. To support this flexibility, this protocol defines that every message must contain a `data-type` field, so a sender can indicate the correct format which the receiver should use to parse a message.
+
+#### Threading
+
+This protocol supports but does not require the use of the `~thread` decorator to provide message threading.
+
+## Reference
+
+#### Message Format
+
+This protocol defines one DIDComm message format, of type `https://didcomm.org/json/1.0/message`. A `message` only requires (in addition to the standard contents of DIDComm messages) a `data-type` field and some number of JSON attachments, under the optional fields `data`, `data~attach`, and `~attach`. 
+
+The `"data-type"` field indicates the type of data contained under the attachments.
+
+The attachment fields are all optional, though there should be at least one.
+
+```json
+{
+  "@id": "123456780",
+  "@type": "https://didcomm.org/json/1.0/message",
```

Comment:
a topic field would be useful as an additional mechanism for recipients to more flexibly register/filter incoming messages.

Reply:
I'm not sure I fully understand what you're after. Are you looking for the ability to specify the `@type` field still with a unique URI then also have the ability to include other details about the specifics of the message so it can be parsed differently?

If so, this could be defined within the data model without concern. If we need to move it to a shared attribute that can't be collided with then I think it would require more input and consideration from others.

Reply:
@kdenhartog my comment was not intended to be about the type field. @Moopli added a purpose field into the data model.

You can also see how DIF Identity Hubs defined Actions (comment below - https://github.com/hyperledger/aries-rfcs/pull/351/files#r362265169)

Reply:
@kdenhartog  Note: from the above there are (at least) three possible variations:
1. `@type`=`https://didcomm.org/json/1.0/message`
  - `data-type` defines the schema of the payload.
  - `purpose` defines the Intent allowing external apps to register.
  - attachment has the payload (and meta info like mime-type).
2. `@type` defines the schema.
  - `~purpose` (as a decorator) defines the Intent allowing external apps to register.
  - attachment has the payload (and meta info like mime-type).
3. `@type` contains `https://didcomm.org/json/1.0/message` and also some schema information.
  - `purpose` defines the Intent allowing external apps to register.
  - attachment has the payload (and meta info like mime-type).

The RFC currently uses the first variation. Between 2 & 3, I would assume the decorator style (2) would be more consistent.